src/mesh/mesh_refinement.cpp

```diff
@@ -303,164 +303,33 @@ void MeshRefinement::ProlongateCellCenteredValues(const ParArrayND<Real> &coarse
-          Real gx1p = (coarse(n, k, j, i + 1) - ccval) / dx1p;
-          Real gx1c =
-              0.5 * (SIGN(gx1m) + SIGN(gx1p)) * std::min(std::abs(gx1m), std::abs(gx1p));
-          Real gx2m = (ccval - coarse(n, k, j - 1, i)) / dx2m;
-          Real gx2p = (coarse(n, k, j + 1, i) - ccval) / dx2p;
-          Real gx2c =
-              0.5 * (SIGN(gx2m) + SIGN(gx2p)) * std::min(std::abs(gx2m), std::abs(gx2p));
-
-          // KGF: add the off-centered quantities first to preserve FP symmetry
-          // interpolate onto the finer grid
-          fine(n, fk, fj, fi) = ccval - (gx1c * dx1fm + gx2c * dx2fm);
-          fine(n, fk, fj, fi + 1) = ccval + (gx1c * dx1fp - gx2c * dx2fm);
-          fine(n, fk, fj + 1, fi) = ccval - (gx1c * dx1fm - gx2c * dx2fp);
-          fine(n, fk, fj + 1, fi + 1) = ccval + (gx1c * dx1fp + gx2c * dx2fp);
-        });
-  } else { // 1D
-    int k = ckb.s, fk = kb.s, j = cjb.s, fj = jb.s;
-    pmb->par_for(
-        "ProlongateCellCenteredValues1d", sn, en, si, ei,
-        KOKKOS_LAMBDA(const int n, const int i) {
-          int fi = (i - cib.s) * 2 + ib.s;
-          const Real x1m = coarse_coords.x1v(i - 1);
-          const Real x1c = coarse_coords.x1v(i);
-          const Real x1p = coarse_coords.x1v(i + 1);
-          Real dx1m = x1c - x1m;
-          Real dx1p = x1p - x1c;
-          const Real fx1m = coords.x1v(fi);
-          const Real fx1p = coords.x1v(fi + 1);
-          Real dx1fm = x1c - fx1m;
-          Real dx1fp = fx1p - x1c;
-
-          Real ccval = coarse(n, k, j, i);
-
-          // calculate 1D gradient using the min-mod limiter
-          Real gx1m = (ccval - coarse(n, k, j, i - 1)) / dx1m;
-          Real gx1p = (coarse(n, k, j, i + 1) - ccval) / dx1p;
-          Real gx1c =
-              0.5 * (SIGN(gx1m) + SIGN(gx1p)) * std::min(std::abs(gx1m), std::abs(gx1p));
-
-          // interpolate on to the finer grid
-          fine(n, fk, fj, fi) = ccval - gx1c * dx1fm;
-          fine(n, fk, fj, fi + 1) = ccval + gx1c * dx1fp;
-        });
+  int b = 0;
+  int nbuffers = coarse.GetDim(6) * coarse.GetDim(5);
+  int nvars = coarse.GetDim(4);
+  cell_centered_bvars::BufferCache_t info("refinement info", nbuffers);
+  auto info_h = Kokkos::create_mirror_view(info);
+  for (int l = 0; l < coarse.GetDim(6); ++l) {
+    for (int m = 0; m < coarse.GetDim(5); ++m) {
+      // buff and var unused
+      info_h(b).si = si;
+      info_h(b).ei = ei;
+      info_h(b).sj = sj;
+      info_h(b).ej = ej;
+      info_h(b).sk = sk;
+      info_h(b).ek = ek;
+      info_h(b).Nt = coarse.GetDim(6);
+      info_h(b).Nu = coarse.GetDim(5);
+      info_h(b).Nv = nvars;
+      info_h(b).refinement_op = RefinementOp_t::Prolongation;
+      info_h(b).coords = pmb->coords;
+      info_h(b).coarse_coords = this->coarse_coords;
+      info_h(b).fine = fine.Get();
+      info_h(b).coarse = coarse.Get();
+      ++b;
+    }
   }
-  return;
+  Kokkos::deep_copy(info, info_h);
```

Comment:
Does this `deep_copy` occur at every step that prolongation occurs on? Is that necessary?

Reply:
Yes. And no respectively. However, I would argue we just leave it for now, assuming it doesn't impact performance too much, as this machinery is per-meshblock, and I would like to remove it eventually.